Libraries/pbxbuild/Sources/Target/Environment.cpp

```diff
@@ -269,8 +304,8 @@ Create(Build::Environment const &buildEnvironment, Build::Context const &buildCo
             projectActionEnvironment.insertFront(level, false);
         }
 
-        projectConfigurationFile = LoadConfigurationFile(projectConfiguration, projectActionEnvironment);
-        if (projectConfigurationFile != nullptr) {
+        projectConfigurationFile = LoadConfigurationFile(Filesystem::GetDefaultUNSAFE(), projectActionEnvironment, workingDirectory, projectConfiguration);
```

Comment:
From 8e36066203c13d0dee7911c8e8bc168eb8d1da67. It's unsafe because calling code can't pass down the filesystem — it always uses a `DefaultFilesystem` (i.e. the real one), even from a unit test using a `MemoryFilesystem`. It can even modify the filesystem without declaring a (non-const) `Filesystem *` parameter in its API.

The issue in this case is that of the main parts of the build:

 1. Load developer directory, platforms, SDKs, toolchains.
 2. Load the workspace and project.
 3. Determine build settings, compute invocations & auxiliary files.
 4. Generate Ninja or run the invocations' commands.

Parts 1, 2, and 4 clearly need filesystem access, but part 3 only occasionally does. My hope is that the filesystem can be removed from part 3 completely, to isolate and test the complex logic in computing settings & invocations. The plan for that is split on reads and writes:

 - Reads get pre-loaded in phase 2, as part of loading the project structure. This is complicated for xcconfigs specifically because build settings (including transient build-specific ones, like command-line build settings) can affect the path the xcconfig is loaded from, and phase 2 is build-independent.
 - Writes get deferred from phase 3 to phase 4, like how auxiliary files are already described in invocations rather than written out directly by the resolving code.